### DISCUSSION TITLE
It's possible to import user from LDAP to Keycloak only the first time the user log in to Keycloak?

### DISCUSSION THREAD

**Discussion by Jackojack7** [keycloak/keycloak#15611] [Q&A]
In the Keycloak [documentation](https://www.keycloak.org/docs/latest/server_admin/#synchronizing-ldap-users-to-keycloak), under Synchronizing LDAP users to Keycloak, it says:

> If you set the Import Users option, the LDAP Provider handles importing LDAP users into the Keycloak local database. The first time a user logs in, the LDAP provider imports the LDAP user into the Keycloak database and validates the LDAP password. This first time a user logs in is the only time Keycloak imports the user. If you click the Users menu in the Admin Console and click the View all users button, you only see the LDAP users authenticated at least once by Keycloak. Keycloak imports users this way, so this operation does not trigger an import of the entire LDAP user database.

But.. actually, I get the exact opposite behavior. When I add LDAP configuration to Keycloak (v.20), it immediately syncs all the LDAP users to my Keycloak users list. How can I import just the users only on the first time a user logs in, as the documentation states.

the configuration I use:

```
{
  providerId: 'ldap',
  providerType: 'org.keycloak.storage.UserStorageProvider',
  name: 'ldap',
  config: {
    enabled: ['true'],
    connectionUrl: [''],
    bindDn: [''],
    bindCredential: [''],
    startTls: ['false'],
    useTruststoreSpi: ['ldapsOnly'],
    connectionPooling: ['false'],
    authType: ['simple'],
    allowKerberosAuthentication: ['false'],
    useKerberosForPasswordAuthentication: ['false'],
    cachePolicy: ['DEFAULT'],
    connectionTimeout: [''],
    usersDn: [''],
    usernameLDAPAttribute: ['sAMAccountName'],
    rdnLDAPAttribute: ['cn'],
    uuidLDAPAttribute: ['objectGUID'],
    userObjectClasses: ['person, organizationalPerson, user'],
    customUserSearchFilter: [''],
    readTimeout: [''],
    batchSizeForSync: [''],
    vendor: ['ad'],
    editMode: ['READ_ONLY'],
    searchScope: [''],
    pagination: ['false'],
    importEnabled: ['true'],
    syncRegistrations: ['false'],
    usePasswordModifyExtendedOp: ['false'],
    validatePasswordPolicy: ['false'],
    trustEmail: ['false'],
    changedSyncPeriod: ['-1'],
    fullSyncPeriod: ['-1'],
  },
```
I have also asked it on - [stackoverflow](https://stackoverflow.com/questions/74515627/can-i-import-users-from-ldap-to-keycloak-only-the-first-time-they-logged-in-to-k)

**Comment by cm253** [comment]
I don't think that's possible with the default LDAP provider from Keycloak. It's designed to synchronize the LDAP users to Keycloak. If the user doesn't exist in Keycloak it's not possible to authenticate/login with it.

What we did is to create our own LDAP provider that only authenticates the user's email and password with LDAP (ldap bind), and not store any user information in Keycloak at all.

**Reply by Jackojack7** [reply]
What you are describing here is the behavior I experience too:

>  If the user doesn't exist in Keycloak it's not possible to authenticate/login with it

However, as you can see, as to the Keycloak documentation I mentioned, another option is available:

> If you set the Import Users option, the LDAP Provider handles importing LDAP users into the Keycloak local database. The first time a user logs in, the LDAP provider imports the LDAP user into the Keycloak database and validates the LDAP password. This first time a user logs in is the only time Keycloak imports the user. If you click the Users menu in the Admin Console and click the View all users button, you only see the LDAP users authenticated at least once by Keycloak. Keycloak imports users this way, so this operation does not trigger an import of the entire LDAP user database.

The question is - how can I achieve it

**Reply by Jackojack7** [reply]
There's any solution here? I would like to sync LDAP user only if he logged in via keycloak
